src/lib/NodeHighlighter.ts

```diff
@@ -0,0 +1,220 @@
+    this.events.on(type, listener);
+  };
+
+  public removeEventListener = <E extends keyof NodeHighlighterMessagesMap>(
+    type: E,
+    listener: NodeHighlighterMessagesMap[E]
+  ): void => {
+    this.events.off(type, listener);
+  };
+
+  public enable = (): void => {
+    if (this.enabled) return;
+
+    this.listenToGlobalEvents();
+    this.observeDomMutations();
+    this.enabled = true;
+  };
+
+  public disable = (): void => {
+    if (!this.enabled) return;
+
+    this.unlistenToGlobalEvents();
+    this.disconnectObservers();
+    this.renderer.clear();
+    this.enabled = false;
+  };
+
+  private highlightVisibleElements = (): void => {
+    this.renderer.render(this.visibleElements);
+  };
+
+  private listenToGlobalEvents = (): void => {
+    window.addEventListener('scroll', this.highlightVisibleElements, { passive: true, capture: true });
+    window.addEventListener('resize', this.highlightVisibleElements, { passive: true });
+    window.addEventListener('click', this.onElementClick, { capture: true });
+  };
+
+  private unlistenToGlobalEvents = (): void => {
+    window.removeEventListener('scroll', this.highlightVisibleElements, { capture: true });
+    window.removeEventListener('resize', this.highlightVisibleElements);
+    window.removeEventListener('click', this.onElementClick, { capture: true });
+  };
+
+  private observeDomMutations = (): void => {
+    if (this.enabled) return;
+
+    document.querySelectorAll(`*[${DataAttribute.ElementCodename}]`).forEach((element: Element) => {
+      if (element instanceof HTMLElement) {
+        this.observeElementVisibility(element);
+      }
+    });
+
+    this.mutationObserver.observe(window.document.body, {
+      childList: true,
+      subtree: true,
+    });
+  };
+
+  private disconnectObservers = (): void => {
+    this.mutationObserver.disconnect();
+    this.intersectionObserver.disconnect();
+    this.observedElements = new WeakSet<HTMLElement>();
+    this.visibleElements = new Set<HTMLElement>();
+  };
+
+  private observeElementVisibility = (element: HTMLElement): void => {
+    if (this.observedElements.has(element)) return;
+
+    this.intersectionObserver.observe(element);
+    this.observedElements.add(element);
+
+    element.addEventListener('mousemove', this.onElementMouseEnter);
+    element.addEventListener('mouseleave', this.onElementMouseLeave);
+  };
+
+  private unobserveElementVisibility = (element: HTMLElement): void => {
+    if (!this.observedElements.has(element)) return;
+
+    this.intersectionObserver.unobserve(element);
+    this.observedElements.delete(element);
+    this.visibleElements.delete(element);
+
+    element.removeEventListener('mousemove', this.onElementMouseEnter);
+    element.removeEventListener('mouseleave', this.onElementMouseLeave);
+  };
+
+  private onDomMutation = (mutations: MutationRecord[]): void => {
+    const attrName = dataToDatasetAttributeName(DataAttribute.ElementCodename);
+    let shouldRerender = false;
```

Comment:
Do not use `let`, you should extract that logic below to a method `shouldRerender` instead (or similar if you need some active behavior)